chuffed/globals/edit_distance.cpp

```diff
@@ -0,0 +1,467 @@
+            min_id_cost = std::min(min_id_cost, deletion_cost[i]);
+        }
+
+        // set maximum possible upper bound in the beginning
+        lastBound = seqSize * 2 * max_id_cost;
+
+        cellChanges = 0;
+        for (int i = 0; i < seqSize * 2; i++) {
+            cellHasChanged[i] = 0;
+        }
+
+
+#ifndef NDEBUG
+        std::cout << "insertion_cost: ";
+        for (int i=0; i<max_char; i++) {
+            std::cout << insertion_cost[i] << " ";
+        }
+        std::cout << std::endl;
+        std::cout << "deletion_cost: ";
+        for (int i=0; i<max_char; i++) {
+            std::cout << deletion_cost[i] << " ";
+        }
+        std::cout << std::endl;
+        std::cout << "substitution_cost: ";
+        for (int i=0; i<max_char*max_char; i++) {
+            std::cout << substitution_cost[i] << " ";
+        }
+        std::cout << std::endl;
+#endif
+
+
+        // we set this propagator to low priority
+        priority = 3;
+
+        // attach variable views
+        int offset = 0;
+
+        for (int i = 0; i < seqSize; i++) seq1[i].attach(this, offset + i, EVENT_C);
+
+        offset += seqSize;
+
+        for (int i = 0; i < seqSize; i++) seq2[i].attach(this, offset + i, EVENT_C);
+
+        // in the future we could also wake the propagator on changes on the edit distance
+        offset += seqSize;
+        ed.attach(this, offset, EVENT_L);
+
+        // insert 0 values into matrices
+        for (int i = 0; i < (seqSize + 1) * (seqSize + 1); i++) {
+            dpMatrix[i] = 0;
+        }
+    }
+
+    void wakeup(int i, int c) override {
+        if (i < seqSize * 2) {
+            if ((unsigned)c & EVENT_C) {
+                if (cellHasChanged[i] == 0) {
+                    cellHasChanged[i] = 1;
+                    cellChanges++;
+                }
+            }
+        }
+
+        pushInQueue();
+    }
+
+    bool propagate() override {
+        // Step 1:
+        // Update dp matrix according to change
+        //
+
+#ifndef  NDEBUG
+        std::cout << "lastBound = " << lastBound << std::endl;
+        std::cout << "cellChanges = " << cellChanges << std::endl;
+        std::cout << "cellHasChanged = [";
+        for (int i = 0; i < seqSize * 2; i++) {
+            std::cout << cellHasChanged[i] << " ";
+        }
+        std::cout << "]" << std::endl;
+#endif
+
+          int ub = std::min(2 * seqSize * max_id_cost, lastBound + cellChanges * 2 * max_id_cost);
+//        int ub = 2 * seqSize * max_id_cost;
+
+        updateDpMatrix(ub);
+        cellChanges = 0;
+        for (int i = 0; i < seqSize * 2; i++) {
+            cellHasChanged[i] = 0;
+        }
+
+
+#ifndef NDEBUG
+        printCurrentDpMatrix();
+#endif
+
+        // Step 2:
+        // Propagate lower bound to edit distance, and add explanation
+        //
+
+        // calc edit distance
+        int editDistanceLB = getEditDistanceLB();
+
+#ifndef NDEBUG
+//        updateDpMatrix(2 * seqSize * max_id_cost);
```

Comment:
If this line and the following ones are not needed then please remove them from the code.